sdk/identity/azure-identity/src/token_credential_impl.cpp

```diff
@@ -87,103 +108,193 @@ Azure::Core::Credentials::AccessToken TokenCredentialImpl::GetToken(
-    if (responseBodyPos == std::string::npos)
-    {
-      std::ostringstream errorMsg;
-      errorMsg << errorMsgPrefix << "response json: \'" << jsonExpiresIn << "\' not found.";
+  int64_t expiresIn = 0;
+  if (GetPropertyValueAsInt64(jsonString, expiresInPropertyName, expiresIn))
+  {
+    accessToken.ExpiresOn = std::chrono::system_clock::now() + std::chrono::seconds(expiresIn);
+    return accessToken;
+  }
 
-      throw AuthenticationException(errorMsg.str());
-    }
+  if (expiresOnPropertyName.empty())
+  {
+    ThrowMissingJsonPropertyError(expiresInPropertyName);
+  }
 
-    for (; responseBodyPos < responseBodySize; ++responseBodyPos)
+  std::string expiresOn;
+  if (!GetPropertyValueAsString(jsonString, expiresOnPropertyName, expiresOn))
+  {
+    ThrowMissingJsonPropertyError(expiresInPropertyName + "\' or \'" + expiresOnPropertyName);
+  }
+
+  {
+    auto const spacePos = expiresOn.find(' ');
+    if (spacePos != std::string::npos) // LCOV_EXCL_LINE
     {
-      auto c = responseBody[responseBodyPos];
-      if (c != ':' && c != ' ' && c != '\"' && c != '\'')
-      {
-        break;
-      }
+      expiresOn = expiresOn.replace(spacePos, 1, 1, 'T');
     }
+  }
+
+  accessToken.ExpiresOn = Azure::DateTime::Parse(expiresOn, Azure::DateTime::DateFormat::Rfc3339);
+  return accessToken;
+}
+
+namespace {
+std::string::size_type GetPropertyValueStart(
+    std::string const& jsonString,
+    std::string const& propertyName);
 
-    long long expiresInSeconds = 0;
-    for (; responseBodyPos < responseBodySize; ++responseBodyPos)
+bool GetPropertyValueAsInt64(
+    std::string const& jsonString,
+    std::string const& propertyName,
+    int64_t& outValue)
+{
+  auto const valueStartPos = GetPropertyValueStart(jsonString, propertyName);
+  if (valueStartPos == std::string::npos)
+  {
+    return false;
+  }
+
+  int64_t value = 0;
+  {
+    auto const size = jsonString.size();
+    for (auto pos = valueStartPos; pos < size; ++pos)
     {
-      auto c = responseBody[responseBodyPos];
+      auto c = jsonString[pos];
       if (c < '0' || c > '9')
       {
         break;
       }
 
-      expiresInSeconds = (expiresInSeconds * 10) + (static_cast<long long>(c) - '0');
+      value = (value * 10) + (static_cast<int64_t>(c) - '0');
     }
+  }
 
-    responseBodyPos = responseBody.find(':', responseBody.find(jsonAccessToken));
-    if (responseBodyPos == std::string::npos)
-    {
-      std::ostringstream errorMsg;
-      errorMsg << errorMsgPrefix << "response json: \'" << jsonAccessToken << "\' not found.";
+  outValue = value;
 
-      throw AuthenticationException(errorMsg.str());
-    }
+  return true;
+}
+
+std::string::size_type GetPropertyValueEnd(std::string const& str, std::string::size_type startPos);
+
+bool GetPropertyValueAsString(
+    std::string const& jsonString,
+    std::string const& propertyName,
+    std::string& outValue)
+{
+  auto const valueStartPos = GetPropertyValueStart(jsonString, propertyName);
```

Comment:
Is the nlohmann JSON package too slow to be used here? I'm a bit concerned about implementing a client-side parser here.

Reply:
I bet that `nlohmann Json` is 100x faster than the `azure cli` :) .  The Azure cli is python, and it will take a couple of seconds to return the token.  The json parsing from the credential would be much faster.
In java, for example, the credential is manually adding a delay of 10secods to wait for the Azure CLI to finish :)

Reply:
At the time this was written, we did not have nlohmann JSON as part of our SDK. This is just moving things around, extracting the code that does parsing to its own method. We can convert to using nlohmann in a separate PR.

Reply:
@vhvb1989 When we build the perf tests for identity, we will be driving the system to the point of CPU exhaustion (because the tests aren't measuring the performance of the SDK if the system isn't 100% CPU bound). So JSON parsing speed can and will become relevant.

Reply:
In a separate PR I'll change it.
But @vhvb1989, I bet the opposite - for this specific case, my code is faster than the general-purpose `nlohmann Json`. It also does not handle many situations. Which is why we should change it and I'll change that in a separate PR.

Reply:
#4155